### PR TITLE
Org.onionshare.onion share

### PR DESCRIPTION
**Pull request by micahflee**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by kushaldas** [comment]
It will be an amazing addition. Thank you @micahflee.

File: README.md [hunk 1]
```diff
@@ -1,27 +1,8 @@
-# Flathub
```
Comment (barthalion) [inline README.md:1]: Feel free to delete this file too.

File: org.onionshare.OnionShare.yaml [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+sdk: org.kde.Sdk
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.golang
+separate-locales: false
+finish-args:
+  - "--device=dri"
+  - "--share=ipc"
+  - "--share=network"
+  - "--socket=wayland"
+  - "--socket=x11"
```
Comment (barthalion) [inline org.onionshare.OnionShare.yaml:15]:
```suggestion
  - "--socket=fallback-x11"
```

File: org.onionshare.OnionShare.yaml [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+  - "--talk-name=org.freedesktop.Flatpak"
+  - "--talk-name=org.freedesktop.Notifications"
+  - "--talk-name=org.freedesktop.secrets"
+  - "--filesystem=home:ro"
+  - "--filesystem=~/OnionShare:create"
+  - "--filesystem=xdg-config/onionshare:create"
```
Comment (barthalion) [inline org.onionshare.OnionShare.yaml:21]:
```suggestion
```

Your application has write access to `$XDG_CONFIG_HOME` by default.

File: org.onionshare.OnionShare.yaml [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.golang
+separate-locales: false
+finish-args:
+  - "--device=dri"
+  - "--share=ipc"
+  - "--share=network"
+  - "--socket=wayland"
+  - "--socket=x11"
+  - "--talk-name=org.freedesktop.Flatpak"
```
Comment (barthalion) [inline org.onionshare.OnionShare.yaml:16]: What do you need it for?

File: org.onionshare.OnionShare.yaml [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+    sources:
+      - type: archive
+        sha256: f175c1d8813257904cf0efeb58e44f68d53b9916f73adaf9ce19514c0271c3fa
+        url: https://download.qt.io/official_releases/QtForPython/pyside2/PySide2-5.15.1-src/pyside-setup-opensource-src-5.15.1.tar.xz
+      - type: shell
+        commands:
+          - mkdir -p /app/include/qt5tmp && cp -R /usr/include/Qt* /app/include/qt5tmp # https://bugreports.qt.io/browse/PYSIDE-787
+          - sed -i 's|\(--include-paths=\)|\1/app/include/qt5tmp:|' sources/pyside2/cmake/Macros/PySideModules.cmake
+  - name: tor
+    buildsystem: simple
```
Comment (barthalion) [inline org.onionshare.OnionShare.yaml:43]: Why not use autotools here?

File: org.onionshare.OnionShare.yaml [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+      - "./configure --prefix=${FLATPAK_DEST}"
+      - make
+      - make install
+    sources:
+      - type: archive
+        sha256: 22cba3794fedd5fa87afc1e512c6ce2c21bc20b4e1c6f8079d832dc1e545e733
+        url: https://dist.torproject.org/tor-0.4.5.6.tar.gz
+    modules:
+      - name: libevent
+        buildsystem: simple
```
Comment (barthalion) [inline org.onionshare.OnionShare.yaml:54]: Same as above.

File: org.onionshare.OnionShare.yaml [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+            url: https://github.com/libevent/libevent/releases/download/release-2.1.12-stable/libevent-2.1.12-stable.tar.gz
+            sha256: 92e6de1be9ec176428fd2367677e61ceffc2ee1cb119035037a27d346b0403bb
+  - name: obfs4proxy
+    buildsystem: simple
+    build-options:
+      env:
+        GOBIN: "/app/bin/"
+        GO111MODULE: "off"
+    build-commands:
+      - ". /usr/lib/sdk/golang/enable.sh; GOPATH=$PWD go install gitlab.com/yawning/obfs4.git/obfs4proxy"
```
Comment (barthalion) [inline org.onionshare.OnionShare.yaml:70]: Does it actually work? Flathub runs builds offline, https://github.com/flatpak/flatpak-builder-tools/tree/master/go-get can be used to generate manifest automatically.

File: org.onionshare.OnionShare.yaml [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+    build-commands:
+      - python3 setup.py install --prefix=${FLATPAK_DEST}
+      - install -D -m0644 org.onionshare.OnionShare.appdata.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.appdata.xml
+      - install -D -m0644 org.onionshare.OnionShare.svg ${FLATPAK_DEST}/share/icons/hicolor/scalable/apps/org.onionshare.OnionShare.svg
+      - install -D -m0644 org.onionshare.OnionShare.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+    sources:
+      - type: dir
+        path: ../desktop/src
+    modules:
+      - name: python3-qrcode
```
Comment (barthalion) [inline org.onionshare.OnionShare.yaml:125]: There's a generator for pip too, if it makes the work simpler for you: https://github.com/flatpak/flatpak-builder-tools/tree/master/pip

**Comment by barthalion** [comment]
bot, build org.onionshare.OnionShare

File: org.onionshare.OnionShare.yaml [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+finish-args:
+  - "--device=dri"
+  - "--share=ipc"
+  - "--share=network"
+  - "--socket=wayland"
+  - "--socket=x11"
+  - "--talk-name=org.freedesktop.Flatpak"
+  - "--talk-name=org.freedesktop.Notifications"
+  - "--talk-name=org.freedesktop.secrets"
+  - "--filesystem=home:ro"
```
Comment (Erick555) [inline org.onionshare.OnionShare.yaml:19]: Can this app use portals? If yes then you may not need this permission.

**Comment by micahflee** [comment]
bot, build org.onionshare.OnionShare

**Comment by micahflee** [comment]
Okay, I've fixed the manifest and addressed all of your comments, and it seems ready to go.

> Your application has write access to `$XDG_CONFIG_HOME` by default.

Without this in `finish-args`:

```
   - "--filesystem=xdg-config/onionshare:create" 
```

I get this error:

```
$ flatpak run org.onionshare.OnionShare
OnionShare 2.3.1 | https://onionshare.org/
Gtk-Message: 19:29:10.970: Failed to load module "canberra-gtk-module"
Gtk-Message: 19:29:10.971: Failed to load module "canberra-gtk-module"
Qt: Session management error: None of the authentication protocols specified are supported
Traceback (most recent call last):
  File "/app/bin/onionshare", line 33, in <module>
  File "/app/lib/python3.8/site-packages/onionshare-2.3.1-py3.8.egg/onionshare/__init__.py", line 177, in main
OSError: [Errno 30] Read-only file system: '/home/user/.config/onionshare/lock'
```

The problem is in upstream OnionShare is writing [directly](https://github.com/micahflee/onionshare/blob/dfaa55d2a40edfb892e123f370f5d22b79e33d04/cli/onionshare_cli/common.py#L141) to `~/.config/onionshare` instead of to `$XDG_CONFIG_HOME` . So I'm keeping it for now, but I opened in issue upstream: https://github.com/micahflee/onionshare/issues/1293

I made the `tor` and `libevent` modules use autotools.

I actually did use the flatpak-builder-tools repo for both pip dependencies and for the golang obfs4proxy module, though it turns out https://github.com/flatpak/flatpak-builder-tools/tree/master/go-get doesn't work anymore as of about 6 days ago when https://github.com/flathub/org.freedesktop.Sdk.Extension.golang updated golang to 1.16 -- it took awhile to realize I needed the `GO111MODULE=off` environment variable for this older way of getting go sources to work.

**Comment by micahflee** [comment]
bot, build org.onionshare.OnionShare

**Comment by micahflee** [comment]
Oh yeah and

> Can this app use portals? If yes then you may not need this permission.

I'd like to use portals, but it's not yet implemented.

**Pull request closed without merging**